pkg/tfshim/walk/walk.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// Drill down into a Map, Set or List element schema.
+type ElementStep struct{}
+
+func (ElementStep) isSchemaPathStep() {}
+
+func (step ElementStep) GoString() string {
+	return "walk.ElementStep{}"
+}
+
+func (step ElementStep) Lookup(s shim.Schema) (shim.Schema, error) {
+	switch elem := s.Elem().(type) {
+	case shim.Resource:
+		switch s.Type() {
+		case shim.TypeMap:
+			return nil, fmt.Errorf("%s is not applicable to object types", step.GoString())
+		case shim.TypeList, shim.TypeSet:
+			return wrapSchemaMap(elem.Schema()), nil
+		default:
+			return nil, fmt.Errorf("%s is not applicable", step.GoString())
+		}
+	case shim.Schema:
+		return elem, nil
+	default:
+		return nil, fmt.Errorf("%s is not applicable", step.GoString())
+	}
+}
+
+func wrapSchemaMap(sm shim.SchemaMap) shim.Schema {
+	return (&schema.Schema{
+		Type: shim.TypeMap,
+		Elem: (&schema.Resource{Schema: sm}).Shim(),
+	}).Shim()
+}
+
+// Utility function to recognize nested object field type schemas encoded in shim.Schema.
+func unwrapSchemaMap(s shim.Schema) (shim.SchemaMap, bool) {
+	switch elem := s.Elem().(type) {
+	case shim.Resource:
+		return elem.Schema(), true
+	default:
+		return nil, false
+	}
+}
+
+type SchemaVisitor = func(SchemaPath, shim.Schema)
+
+// Visit all nested schemas, including the current one.
+func VisitSchema(schema shim.Schema, visitor SchemaVisitor) {
+	visitSchemaInner(NewSchemaPath(), schema, visitor)
+}
+
+// Visit all nested schemas in a SchemaMap, keeping track of SchemaPath location.
+func VisitSchemaMap(schemaMap shim.SchemaMap, visitor SchemaVisitor) {
+	visitSchemaMapInner(NewSchemaPath(), schemaMap, visitor)
+}
+
+func visitSchemaInner(path SchemaPath, schema shim.Schema, visitor SchemaVisitor) {
+	visitor(path, schema)
+	switch elem := schema.Elem().(type) {
+	case shim.Resource:
+		var nestedPath SchemaPath
+		if schema.Type() == shim.TypeMap {
+			// Single-nested blocks are special, drilling down into the elements of the block's object type
+			// can begin immediately without an Element step.
+			nestedPath = path
+		} else {
+			nestedPath = path.Element()
+		}
+		visitSchemaMapInner(nestedPath, elem.Schema(), visitor)
+	case shim.Schema:
+		visitSchemaInner(path.Element(), elem, visitor)
+	}
+}
+
+func visitSchemaMapInner(path SchemaPath, schemaMap shim.SchemaMap, visitor SchemaVisitor) {
+	schemaMap.Range(func(key string, schema shim.Schema) bool {
+		visitSchemaInner(path.GetAttr(key), schema, visitor)
+		return true
+	})
+}
+
+// Converts a value path to a Schema Path (hashicorp/go-cty representation).
+func FromHCtyPath(path cty.Path) SchemaPath {
+	p := NewSchemaPath()
+	for _, subPath := range path {
+		switch s := subPath.(type) {
+		case cty.IndexStep:
+			p = p.Element()
+		case cty.GetAttrStep:
+			p = p.GetAttr(s.Name)
+		}
+	}
+	return p
+}
+
+// Converts a value path to a Schema Path (zclconf package representation).
+func FromCtyPath(path hcty.Path) SchemaPath {
```

Comment:
I think the names on these functions is backwards. 
```suggestion
func FromHCtyPath(path hcty.Path) SchemaPath {
```